Docs: Add technical overview (#563)
steelthreads/agds-next: guides/technical-overview.mdx

---
title: Technical overview
description: Explaining technical decisions in the AgDS component library
---

## React

The AgDS component library has been built using [React](https://reactjs.org/). React is a JavaScript library for building user interfaces and uses a component based model for defining user interface elements.

## NPM Packages

The delivery method for the AgDS component library is via a single [NPM](https://www.npmjs.com/) package. NPM is the package manager for JavaScript and the world's largest software registry.

Every component in the AgDS component library has an individual entry point in the `@ag.ds-next/react` package. This keeps application bundles small and avoids end users downloading unnecessary dependencies.

## TypeScript

All of the AgDS components are built using [Typescript](https://www.typescriptlang.org/), a typed superset of JavaScript that compiles to plain JavaScript.

> You don't need to using Typescript in order to use the AgDS componentg library, but it is highly recommended.

Reviewer: ```suggestion
You don't need to be using Typescript in order to use the AgDS component library, but it is highly recommended.
```

---
order: 5
title: Technical Overview
description: Explaining technical decisions in the AgDS Component library
---

## React

> The AgDS Component library is built using React. React is a JavaScript library for building user interfaces. It is maintained by Facebook and a community of individual developers and companies.

Reviewer 2: ```suggestion
The AgDS Component library is built using [React](https://reactjs.org/). React is a JavaScript library for building user interfaces. It is maintained by Facebook and a community of individual developers and companies.
```

React uses a component-based model for defining UI elements, which is fantastic for building Design Systems and scalable UI generally.

## NPM Packages

> The delivery method for the AgDS Component library is via NPM packages. NPM is the package manager for JavaScript and the world’s largest software registry.

Reviewer 2: ```suggestion
The delivery method for the AgDS Component library is via [NPM](https://www.npmjs.com/) packages. NPM is the package manager for JavaScript and the world’s largest software registry.
```

Each component is an individual package. This allows consumers to only install the components they need, and avoid installing unnecessary dependencies and huge bundles for users.

> ```zsh

Reviewer 2: I don't think this example is necessary. Or if you think it is, can you add a H3 above it?

yarn add @agds-next/button
```

The trade off is that dependency versions need to be managed carefully. Our [release notes](https://steelthreads.github.io/agds-next/releases) include a list of the latest versions of packages, and any breaking changes that need to be managed with a release.

## TypeScript

> All of the AgDS components are built using Typescript, a typed superset of JavaScript that compiles to plain JavaScript. Typing component props is a great way of documenting them, and ensure they are easy to use and free of bugs.

Reviewer 2: ```suggestion
All of the AgDS components are built using [Typescript](https://www.typescriptlang.org/), a typed superset of JavaScript that compiles to plain JavaScript.
```

I'm not sure what the second sentence should say right now, but we shouldn't say our components are free or bugs .

You don't need to use Typescript to use AgDS components, but it is recommended.

## Emotion

> Emotion is a framework that enables writing styling (CSS) within JavaScript components. As opposed to writing CSS stylesheets, the style code is written alongside the component code, making it easier to maintain and reason.

Reviewer 2: ```suggestion
[Emotion](https://emotion.sh/docs/introduction) is a library that enables writing CSS styles within JavaScript files. As opposed to writing separate CSS stylesheets, the CSS is written alongside the component code, making it easier to maintain and reason.
```

## Next.JS

> Next.JS is a popular framework for building React-based web applications. Describes as 'the SDK for the web', it handles the tooling and configuration needed for a React app, and provides additional structure, features, and optimizations for your application. It also has fantastic documentation and developer experience.

Reviewer 2: ```suggestion
[Next.JS](https://nextjs.org/) is a popular framework for building React-based web applications. Describes as 'the SDK for the web', it handles the tooling and configuration needed for a React app, and provides additional structure, features, and optimizations for your application. It also has fantastic documentation and developer experience.
```

We use Next.JS to build our documentation site, templates, and internal test applications.

> Using Next.JS massively improves our productivity, because it has built-in solutions for routing, static and server rendering, API routes and more. That means we spend less time on configuration and we can focus on building application logic and layouts.

Reviewer 2: ```suggestion
Using Next.JS massively improves our productivity due  to it's built-in solutions for routing, static and server rendering, API routes and more. That means we spend less time on configuration and we can focus on building application logic and layouts.
```
